### PR TITLE
feat(diagnostic): add option to include diagnostic source

### LINKED ISSUE #15716: diagnostic: add ability to display diagnostic source in virtual text/show_{line,position}_diagnostics
### Feature description

An extremely popular override in the lsp user community (especially those using null-ls/nvim-lint) is to override the display of the virtual text/show_X_ diagnostic functions in order to also display the diagnostic source. Now that we have an official diagnostic API, we should add this as a configurable option.

@mfussenegger suggested `include_source` with `never`, `always`, and `if_many`

See:
https://github.com/neovim/nvim-lspconfig/wiki/UI-customization#show-source-in-diagnostics-neovim-05106-only

### PR DESCRIPTION
**Pull request by gpanders**
Add an option to virtual text display and floating window previews to include diagnostic source in the diagnostic message.

Closes #15716.

### CODE REVIEW

**Comment by gpanders** [comment]
Marking this as draft until I write a test for it, otherwise it should be good to go (pending review).

**Review by mjlbach — APPROVED** [review]
Approving because the implementation looks correct, however, while I agree with the naming of the options `include_source -> always/if_many` it should probably be discussed.

**Comment by entropitor** [comment]
It would be nice if the user can just provide a function to render a message given a diagnostic. E.g. for eslint it also provides a code which is the rule that is violated and it could be nice to also include the ruleName in the error message.

**Comment by gpanders** [comment]
> It would be nice if the user can just provide a function to render a message given a diagnostic. E.g. for eslint it also provides a code which is the rule that is violated and it could be nice to also include the ruleName in the error message.

There's no reason this isn't already possible. When the list of diagnostics is populated to pass to `vim.diagnostic.set()`, simply add the code to the message.

**Comment by mjlbach** [comment]
@gpanders we shouldn't modify the diagnostics passed to `vim.diagnostic.set()` as it will break code actions if anything is injected into the diagnostic message (there was a convo on the lspconfig issue tracker about this)

**Comment by gpanders** [comment]
We are copying before modifying: https://github.com/neovim/neovim/pull/15717/files#diff-212294e88097d8dc45a881c6c3c5fd4ed266040e49fb07317005969849a70156R67-R69

**Comment by neovim-discourse** [comment]
This pull request has been mentioned on **Neovim Discourse**. There might be relevant details there:

https://neovim.discourse.group/t/diagnostic-source-no-longer-exists-after-vim-diagnostic-overhaul/1151/2

File: runtime/doc/diagnostic.txt [hunk 1]
```diff
@@ -226,6 +227,9 @@ config({opts}, {namespace})                          *vim.diagnostic.config()*
                                    • severity: Only show virtual text for
                                      diagnostics matching the given severity
                                      |diagnostic-severity|
+                                   • include_source: (string) Include the
```
Comment (justinmk) [inline runtime/doc/diagnostic.txt:230]: i would drop "include_" prefix. don't want to set a precedent for that. 
```suggestion
                                   • source: (string) Include the
```

**Review by justinmk — APPROVED** [review]

**Comment by mjlbach** [comment]
Can you squash all of the renaming option commits and finalize the commit message? Otherwise I'm good to merge

**Comment by gpanders** [comment]
Done. I kept the commit removing `get_virt_text_chunks()` so that we have the context for why it was removed in the commit message.

**Comment by kristijanhusak** [comment]
It's a bit of a bummer for me that this option got removed. I have [this setup](https://github.com/kristijanhusak/neovim-config/blob/master/nvim/lua/partials/lsp.lua#L169-L180) to show virtual text only on the current line (Ale did it like that back in the day). Now I'm not really able to format the diagnostics to a proper highlight/message to show it. 
Is there a chance to expose a `vim.diagnostic.config` option to show the diagnostics only on the current line?

**Comment by gpanders** [comment]
The reason `get_virt_text_chunks()` was removed was because it was impossible to provide a coherent interface with the `'source'` option. If a user has `'source'` set to `"if_many"`, that setting would not behave the same in `get_virt_text_chunks()` as it would elsewhere, which is confusing and inconsistent. The alternative is to just not include the source in the virtual text returned from `get_virt_text_chunks()`, but that is also confusing and inconsistent.

It was safe to remove because (as you've discovered) it's easy enough to do without it now. I do the same thing you do: only showing virtual text on the current line. If you can stomach a bit of Lisp, this is what I use:

```fennel
  (local ns (vim.api.nvim_create_namespace :diagnostics))
  (fn print-diagnostics []
    (let [bufnr (vim.api.nvim_get_current_buf)
          [lnum _] (vim.api.nvim_win_get_cursor 0)
          lnum (- lnum 1)
          diagnostics (vim.diagnostic.get bufnr {: lnum})]
      (vim.diagnostic.show ns bufnr diagnostics {:virtual_text true})))
  (autocmd diagnostics [:CursorMoved :CursorHold :InsertLeave] "*" (print-diagnostics))))
```

> Edit: I managed to set it up like this. Setting custom namespace via config for some reason didn't work out as expected (signs were not properly cleared), so I had to grab the namespace from the diagnostic itself.

I would recommend (1) using your own namespace to avoid conflicting with existing diagnostics and (2) passing the `opts` table to `show()` directly (see my example).

**Comment by kristijanhusak** [comment]
>  If you can stomach a bit of Lisp


Jokes aside, I tried using my own namespace, but wasn't able to clear the namespace on all other lines except the current one for some reason. Maybe I'm overthinking it. I'll give your code a try.

**Comment by kristijanhusak** [comment]
This actually works really well. Thanks!

**Pull request merged**